### PR TITLE
[BLOG] Array API support in scikit-learn

### PR DESCRIPTION
**Pull request by thomasjpfan**
## Text styling

- [x] The blog is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [x] If there are headers, they use the proper header tags in order to do so (with only one level-one header).
- [x] All links describe where they link to (for example, check the [Quansight labs website](https://labs.quansight.org/)).
- [x] Any kind of styling that the author uses (for example, bold for emphasis) is consistent throughout the blog.

## Non-text contents

- [x] Blog post featured image is in PNG or JPEG format, **not** SVG.
- [x] All content is represented as text (for example, images need alt text and videos need captions or descriptive transcripts).
- [x] If there are emojis, there are not more than three in a row.
- [x] Don't use [flashing gifs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [x] If it were to be read as plain text, the blog still makes sense and no information is missing.

### CODE REVIEW

File: apps/labs/public/posts/array-api-support-scikit-learn/data-api-scikit-learn-logos-hero.png
Comment (kgryte) [inline apps/labs/public/posts/array-api-support-scikit-learn/data-api-scikit-learn-logos-hero.png:1]: I believe the Data APIs logo is off here and in the other PNG.

File: apps/labs/posts/array-api-support-scikit-learn.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+```python
+def _nanmin(X, axis=None):
+    xp = get_array_namespace(X)
+    if _is_numpy_namespace(xp):
+        return xp.asarray(numpy.nanmin(X, axis=axis))
+    else:
+        # Implements nanmin in terms of the Array API standard
+        mask = xp.isnan(X)
+        X = xp.min(..., X), axis=axis)
```
Comment (kgryte) [inline apps/labs/posts/array-api-support-scikit-learn.md:100]: Maybe I am missing something here, but how is `mask` being used in the current snippet?

Comment (thomasjpfan) [reply apps/labs/posts/array-api-support-scikit-learn.md]: I did not want to complicate the snippet more with the implementation details. But it is a good point to at least use the `mask`. I update the PR: [`811958c` (#765)](https://github.com/Quansight/Quansight-website/pull/765/commits/811958c988c6a66202a310c5cff006de0a2f132b) with your suggestion.

File: apps/labs/public/posts/array-api-support-scikit-learn/data-api-scikit-learn-logos-hero.png
Comment (thomasjpfan) [reply apps/labs/public/posts/array-api-support-scikit-learn/data-api-scikit-learn-logos-hero.png:1]: What is off about the logo? I took the logo from: https://github.com/data-apis/data-apis.github.io/blob/main/assets/theme_logo_description.pdf

In this PR, the logo looks like:

Screenshot 2023-09-05 at 2 58 37 PM

Comment (kgryte) [reply apps/labs/public/posts/array-api-support-scikit-learn/data-api-scikit-learn-logos-hero.png:1]: The "s" is missing from **DataAPIs.**

Comment (kgryte) [reply apps/labs/public/posts/array-api-support-scikit-learn/data-api-scikit-learn-logos-hero.png:1]: I checked. It looks like those assets are also missing the "s". Here is the original issue in which the logo was decided: https://github.com/data-apis/workgroup/issues/26

If you look on the website https://data-apis.org, the "s" is also included. I don't think we've ever referred to the consortium as "Data API", but rather "Data APIs".

Comment (thomasjpfan) [reply apps/labs/public/posts/array-api-support-scikit-learn/data-api-scikit-learn-logos-hero.png:1]: I updated the PR to use the logo from the [annual report](https://github.com/data-apis/data-apis.github.io/blob/main/static/files/2021_annual_report_DataAPIs_Consortium.pdf), which includes the "s".

File: apps/labs/posts/array-api-support-scikit-learn.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
+---
+title: 'Array API Support in scikit-learn'
+published: September 5, 2023
+author: thomas-fan
+description: 'In this blog post, we share how scikit-learn enabled support for the Array API Standard.'
+category: [Array API]
+featuredImage:
+  src: /posts/array-api-support-scikit-learn/data-api-scikit-learn-logos.png
+  alt: 'The Data API logo next to the scikit-learn logo.'
```
Comment (rgommers) [inline apps/labs/posts/array-api-support-scikit-learn.md:9]: minor: API -> APIs (and also 3 lines below)

File: apps/labs/posts/array-api-support-scikit-learn.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+hero:
+  src: /posts/array-api-support-scikit-learn/data-api-scikit-learn-logos-hero.png
+  alt: 'The Data API logo next to the scikit-learn logo.'
+---
+
+The Consortium for Python Data API Standards developed the Array API standard, which aims to define consistent behavior between the ecosystem of array libraries, such as PyTorch, NumPy, and CuPy. The Array API standard enables libraries, such as scikit-learn, to write code once with the standard and have it work on multiple array libraries. With PyTorch tensors or CuPy tensors, it is now possible to run computation on accelerators, such as GPUs.
```
Comment (rgommers) [inline apps/labs/posts/array-api-support-scikit-learn.md:15]: minor textual:
- CuPy tensors -> CuPy arrays
- computation -> computions

File: apps/labs/posts/array-api-support-scikit-learn.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+With the release of scikit-learn 1.3, we enabled experimental Array API for a limited set of machine learning models. Array API support is gradually expanding to include more machine learning models and functionality on the development branch. Scikit-learn depends on the `array_api_compat` library for Array API support. The `array_api_compat` extends the Array API standard to NumPy's, CuPy arrays, and PyTorch Tensors. In this blog post, we cover scikit-learn's public interface for enabling Array API, the performance gain of running on an accelerator, and the challenges we faced when integrating Array API.
```
Comment (rgommers) [inline apps/labs/posts/array-api-support-scikit-learn.md:17]: textual:
- insert "usage" or "support" after "Array API" in the first sentence
- remove "The" from the start of the 4th sentence
- Tensors -> tensors
- The 4th sentence is a little off. Maybe it aimed to say that it extends the _main namespaces_ of NumPy, CuPy and PyTorch?

File: apps/labs/posts/array-api-support-scikit-learn.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+## Benchmarks
```
Comment (rgommers) [inline apps/labs/posts/array-api-support-scikit-learn.md:19]: nit: blank line here

File: apps/labs/posts/array-api-support-scikit-learn.md [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+Scikit-learn was initially developed to run on CPUs with NumPy arrays. With Array API support, a limited set of scikit-learn models and tools can now run with other array libraries and devices like GPUs. The following benchmark results are from running scikit-learn's LinearDiscurementAnalysis with different array libraries:
```
Comment (rgommers) [inline apps/labs/posts/array-api-support-scikit-learn.md:20]: typo in `LinearDiscurementAnalysis`, and it should be typeset as code I think

File: apps/labs/posts/array-api-support-scikit-learn.md [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+![Bar plot with benchmark results comparing NumPy, PyTorch CPUs, and PyTorch GPUs running Linear Discriminant Analysis. The PyTorch GPU results are marked as 27 times faster compared to NumPy for fitting the model and 28 times faster compared to NumPy for prediction.](/posts/array-api-support-scikit-learn/benchmark_results.png)
+
+The training and prediction times are improved when using PyTorch compared to NumPy. Running the computation on PyTorch CPU tensors is faster than NumPy because PyTorch CPU operations are multi-threaded by default.
+
+## scikit-learn's Array API interface
+
+Scikit-learn extended its experimental Array API support in version 1.3 to support NumPy's `ndarrays`, CuPy's `ndararys`, and PyTorch `Tensors`. By themselves, these array objects do not implement the Array API specification. To overcome this limitation, Quansight engineer Aaron Meurer, led the development of `array_api_compat` to adopt the Array API to NumPy, CuPy, and PyTorch. Scikit-learn directly uses `array_api_compat` for its Array API support. There are two APIs enabling Array API in scikit-learn: a global configuration and a context manager. The following example use a context manager:
```
Comment (rgommers) [inline apps/labs/posts/array-api-support-scikit-learn.md:28]: use -> uses

File: apps/labs/posts/array-api-support-scikit-learn.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+```python
+import sklearn
+import torch
+from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
+
+X_torch_cpu, y_torch_cpu = torch.asarray(X_np), torch.asarray(y_np)
```
Comment (rgommers) [inline apps/labs/posts/array-api-support-scikit-learn.md:35]: `X_np` and `y_np` aren't defined, so this code doesn't run. Importing some example dataset would be good.

File: apps/labs/posts/array-api-support-scikit-learn.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+## Conclusion
+
+In recent years, there has been increase usage of accelerators for computation in many domains. The array API standard gives Python libraries, such as scikit-learn, access these accelerators with the same source code. Depending on your code, there are various challenges for adopting Array API, but there are performance and compatibilities benefits for using the API. If you observe any limitations, you are welcome to open issues on their issue tracker. For more information about Array API, you may watch Aaron's [SciPy presentation](https://www.youtube.com/watch?v=16rB-fosAWw), read the SciPy proceedings [paper](https://conference.scipy.org/proceedings/scipy2023/aaron_meurer.html), or read the [Array API documentation](https://data-apis.org/array-api/latest/).
+
+This work has made possible by Quansight's collaboration with Meta's PyTorch team. I'll like to thank Aaron Meurer, Matthew Barber, and Ralf Gommers for the development of `array_api_compat`, which was a vital part of this project's success.
```
[the remark this answers is not in the document]
Comment (rgommers) [reply apps/labs/posts/array-api-support-scikit-learn.md:144]: Then, I'd suggest also acknowledging Olivier and Tim for their review and help with this project, and continuing to push forward on expanding support. Maybe saying something more generic and future-looking here about where this is hopefully heading would be useful.

**Review by rgommers — COMMENTED** [review]
Thanks for this post Thomas! Looks like a good start. I did an initial review. The structure of the story looks good to me.

File: apps/labs/posts/array-api-support-scikit-learn.md [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+import torch
+
+X_np, y_np = make_classification(random_state=0, n_samples=500_000, n_features=300)
+X_torch_cpu, y_torch_cpu = torch.asarray(X_np), torch.asarray(y_np)
+
+with sklearn.config_context(array_api_dispatch=True):
+    lda_torch_cpu = LinearDiscriminantAnalysis()
+    lda_torch_cpu.fit(X_torch_cpu, y_torch_cpu)
+
+    predictions = lda_torch_cpu.predict(X_torch_cpu)
```
Comment (rgommers) [inline apps/labs/posts/array-api-support-scikit-learn.md:44]: This example runs now, but it was slightly annoying that it required installing `array-api-compat` from its `main` branch (it needed `xp.take` to have the `axis` keyword be optional). I'll ask Aaron to cut a new release, so that things work out of the box with a `pip install array-api-compat`.

Comment (rgommers) [reply apps/labs/posts/array-api-support-scikit-learn.md:44]: Actually, let's do that here. @asmeurer would you mind cutting a new release of `array-api-compat`?

File: apps/labs/posts/array-api-support-scikit-learn.md [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+## Benchmarks
+
+Scikit-learn was initially developed to run on CPUs with NumPy arrays. With Array API support, a limited set of scikit-learn models and tools can now run with other array libraries and devices like GPUs. The following benchmark results are from running scikit-learn's `LinearDiscriminantAnalysis` with different array libraries:
+
+![Bar plot with benchmark results comparing NumPy, PyTorch on CPU, and PyTorch on NVIDIA 3090 running Linear Discriminant Analysis. The PyTorch GPU results are marked as 27 times faster compared to NumPy for fitting the model and 28 times faster compared to NumPy for prediction.](/posts/array-api-support-scikit-learn/benchmark_results.png)
```
Comment (rgommers) [inline apps/labs/posts/array-api-support-scikit-learn.md:23]: 3090 GTX or RTX?

**Comment by lezcano** [comment]
When do we plan to have this post up? Would it be possible to publish it this week / early next week. This way, I could cite it in the post I'm writing for the PyTorch blog.

**Comment by rgommers** [comment]
> When do we plan to have this post up? Would it be possible to publish it this week / early next week. This way, I could cite it in the post I'm writing for the PyTorch blog.

I'll pencil in doing a final pass tomorrow. Should be feasible to do early next week.

File: apps/labs/posts/array-api-support-scikit-learn.md (hunk 10 above)
Comment (rgommers) [reply apps/labs/posts/array-api-support-scikit-learn.md:44]: @asmeurer could you upload a release by hand if the auto-publish thing is too messy? Should take only a few minutes for a pure Python package I'd think. It'd be great if this release was up by tomorrow.

**Comment by pavithraes** [comment]
We can publish this with the website release on Tuesday. :)

File: apps/labs/posts/array-api-support-scikit-learn.md [hunk 12]
```diff
@@ -0,0 +1,149 @@
+---
+title: 'Array API Support in scikit-learn'
+published: September 5, 2023
+author: thomas-fan
```
Comment (pavithraes) [inline apps/labs/posts/array-api-support-scikit-learn.md:4]:
```suggestion
author: thomas-j-fan
```

This is the handle we use in Storyblok, and should fix the Vercel deployment CI action. :)

File: apps/labs/posts/array-api-support-scikit-learn.md (hunk 11 above)
Comment (pavithraes) [reply apps/labs/posts/array-api-support-scikit-learn.md:23]: Also, the fact that this is running on a 3090 isn't mentioned in the image, so I suggest we actually note it in the text outside alt-text as well if possible -- perhaps in the previous sentence on line 21. :)

(Also, generally, thanks for writing an excellent alt-text here! ✨)

**Comment by lezcano** [comment]
Tuesday sounds great!

File: apps/labs/posts/array-api-support-scikit-learn.md (hunk 10 above)
Comment (asmeurer) [reply apps/labs/posts/array-api-support-scikit-learn.md:44]: The release is out. I don't know if you need conda-forge too, but that should be out as soon as the bot catches it.

Comment (rgommers) [reply apps/labs/posts/array-api-support-scikit-learn.md:44]: Thanks Aaron! conda-forge would be nice, but less urgent. let's let the bot do its thing

**Comment by rgommers** [comment]
Thanks for the review @pavithraes! Public release on Tuesday sounds good to me too. Also, can you look at the banner image config please? It's missing and I'm not sure what the preferred one to use is.

@thomasjpfan I pushed some copy-edits. Could you check those and address the last open review comment?

**Comment by thomasjpfan** [comment]
Your changes in https://github.com/Quansight/Quansight-website/pull/765/commits/cd4f8ddf80ae99fedf8f1015a063b125c4a706f9 looks okay to me. I pushed https://github.com/Quansight/Quansight-website/pull/765/commits/dce937ffdce59e83c4a612db3959162ecfd7a7ac to address the open review comment.

**Review by rgommers — APPROVED** [review]
LGTM now, and the last commit fixed the hero image. So let's give this a go. Thanks Thomas & reviewers!

**Comment by pavithraes** [comment]
@lezcano - We had a delay in release, post is published now: https://labs.quansight.org/blog/array-api-support-scikit-learn :)
